platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/paos/PaosInInterceptor.java

```diff
@@ -0,0 +1,399 @@
+                InputStream userTokenStream = PaosInInterceptor.class.getResourceAsStream(
+                        "/templates/username.handlebars")) {
+            soapMessage = IOUtils.toString(soapMessageStream);
+            soapfaultMessage = IOUtils.toString(soapfaultMessageStream);
+            securityHeader = IOUtils.toString(securityHeaderStream);
+            usernameToken = IOUtils.toString(userTokenStream);
+        } catch (IOException e) {
+            LOGGER.info("Unable to load templates for PAOS");
+        }
+
+    }
+
+    @Override
+    public void handleMessage(Message message) throws Fault {
+        List authHeader = (List) ((Map) message.getExchange()
+                .getOutMessage()
+                .get(Message.PROTOCOL_HEADERS)).get("Authorization");
+        String authorization = null;
+        if (authHeader != null && authHeader.size() > 0) {
+            authorization = (String) authHeader.get(0);
+        }
+        InputStream content = message.getContent(InputStream.class);
+        String contentType = (String) message.get(Message.CONTENT_TYPE);
+        if (contentType != null && contentType.contains("application/vnd.paos+xml")) {
+            try {
+                SOAPPart soapMessage = SamlProtocol.parseSoapMessage(IOUtils.toString(content));
+                Iterator iterator = soapMessage.getEnvelope()
+                        .getHeader()
+                        .examineAllHeaderElements();
+                IDPEntry idpEntry = null;
+                String relayState = "";
+                String responseConsumerURL = "";
+                String messageId = "";
+                while (iterator.hasNext()) {
+                    Element soapHeaderElement = (SOAPHeaderElement) iterator.next();
+                    if ("RelayState".equals(soapHeaderElement.getLocalName())) {
+                        relayState = DOM2Writer.nodeToString(soapHeaderElement);
+                    } else if ("Request".equals(soapHeaderElement.getLocalName())
+                            && soapHeaderElement.getNamespaceURI()
+                            .equals("urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp")) {
+                        try {
+                            soapHeaderElement = SamlProtocol.convertDomImplementation(soapHeaderElement);
+                            Request ecpRequest = (Request) OpenSAMLUtil.fromDom(soapHeaderElement);
+                            IDPList idpList = ecpRequest.getIDPList();
+                            List<IDPEntry> idpEntrys = idpList.getIDPEntrys();
+                            //choose the right entry, probably need to do something better than select the first one
+                            //but the spec doesn't specify how this is supposed to be done
+                            idpEntry = idpEntrys.get(0);
+                        } catch (WSSecurityException e) {
+                            //TODO figure out IdP alternatively
+                            LOGGER.info(
```

Comment:
Going to leave it as info because that would mean that the IDP metadata is completely wrong and everything is going to fail.